akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -84,6 +88,65 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
       }.to(Sink.ignore).run()
   }
 
+  private def handleUpgradeRequests(
+    handler:     HttpRequest => Future[HttpResponse],
+    settings:    ServerSettings,
+    parallelism: Int,
+    log:         LoggingAdapter
+  ): HttpRequest => Future[HttpResponse] = { req =>
+    req.header[Upgrade] match {
+      case Some(upgrade) if upgrade.protocols.exists(_.name equalsIgnoreCase "h2c") =>
+
+        log.debug("Got h2c upgrade request from HTTP/1.1 to HTTP2")
+
+        // https://http2.github.io/http2-spec/#Http2SettingsHeader 3.2.1 HTTP2-Settings Header Field
+        val upgradeSettings = req.headers.collect {
+          case raw: RawHeader if raw.lowercaseName == Http2SettingsHeader.name =>
+            Http2SettingsHeader.parse(raw.value)
+        }
+
+        upgradeSettings match {
+          // Must be exactly one
+          case immutable.Seq(Success(settingsFromHeader)) =>
+            // inject the actual upgrade request with a stream identifier of 1
+            // https://http2.github.io/http2-spec/#rfc.section.3.2
+            val injectedRequest = Source.single(req.addHeader(Http2StreamIdHeader(1)))
+
+            val serverLayer: Flow[ByteString, ByteString, Future[Done]] = Flow.fromGraph(
+              Flow[HttpRequest]
+                .watchTermination()(Keep.right)
+                .merge(injectedRequest)
```

Comment:
Should it be `prepend` instead? On the other hand, the order shouldn't matter, so `merge` should be fine as well.